### PR TITLE
Agg: Remove 16-bit limits

### LINKED ISSUE #23826: [Bug]: Overflow of 16-bit integer in Agg renderer causes PolyCollections to be drawn at incorrect locations
### Bug summary

Because the pixel x coordinates of scanline cells are casted from 32-bit integers to 16-bit ones in Agg ([this line](https://github.com/matplotlib/matplotlib/blob/3b79f63980c767fabc59ec5ba267818d8b4b5671/extern/agg24-svn/include/agg_scanline_p.h#L96)), some polygons that are outside the drawing area may actually be drawn at an incorrect location if they are far enough from the viewport so that the 16-bit integer used for actual drawing overflows and the value happens to be inside the drawing area.

### Code for reproduction

```python
import matplotlib.pyplot as plt
import matplotlib
matplotlib.use('TkAgg')

# Create a 120x120 px figure
fig = plt.figure(figsize=(120, 120), dpi=1)
# Add axis that fills the whole figure
ax = fig.add_axes([0, 0, 1, 1])
ax.axis('off')
# Plot one fill_between at x, y = 0, width = 100 and height = 50
ax.fill_between([0, 100], 0, 50, color='r')
# Plot another fill_between at x = 65536, y = 50, width=100 and height = 50
# 65536 causes the 16-bit integer to overflow so that the coordinate becomes actually 0
ax.fill_between([2 ** 16, 2 ** 16 + 100], 50, 100, color='b')
# Set limits to ensure that one pixel in the rasterized image corresponds to one unit in the axis
ax.set_xlim(-10, 110)
ax.set_ylim(-10, 110)
# Agg has the bug, Pdf does not
plt.savefig('bug.png', dpi=1)
plt.savefig('nobug.pdf', dpi=1)
plt.show()
```

### Actual outcome

The PolyCollections are both drawn in the image, even though they are 65536 units apart and the x axis goes from -10 to 110.
bug

### Expected outcome

There should only be the red rectangle, as in this PDF.
[nobug.pdf](https://github.com/matplotlib/matplotlib/files/9525611/nobug.pdf)

### Additional information

I debugged this quite a bit, and turns out it is a bit tricky to replicate this. The PolyCollections need to be able to fit in the axes completely so that the optimized `draw_markers` function [is called](https://github.com/matplotlib/matplotlib/blob/3b79f63980c767fabc59ec5ba267818d8b4b5671/lib/matplotlib/collections.py#L406). Then, inside Agg, transforms are applied so that pixel coordinates are calculated. These coordinates are fine, as they use 32-bit integers. However, when Agg draws the scanlines, it converts the coordinates to 16-bit integers. If the axis is zoomed in enough and everything lines up just right, some points that have very high positive or very high negative pixel coordinates may appear in the viewport because the int16 conversion causes the 16-bit coordinate to over- or underflow and to be inside the viewport. An obvious fix could be to use int32 instead of int16 while drawing, but I don't know if that's possible or if matplotlib is even the right place to report, as this bug seems to be in Agg code. The original figure where I discovered this had many fill_betweens, and when zoomed in, some fill_betweens from earlier or later in the data would appear.

Note that at least on my machine, the figure that is shown in the window might not have the bug present, but the saved PNG file should have it.

I checked also the main branch (version `3.6.0.dev3132+gf8cf0ee5e7`), and it has the bug.

### Operating system

Probably anything where Agg runs (verified on RHEL8 and Windows)

### Matplotlib Version

3.5.3

### Matplotlib Backend

QtAgg

### Python version

3.9.13

### Jupyter version

_No response_

### Installation

conda

### LINKED ISSUE #26368: [Bug]: Long audio files result in incomplete spectrogram visualizations
### Bug summary

When using matplotlib to generate spectrogram visualizations of audio files, if the audio file is too long, the spectrogram portion of the plot becomes blank towards the latter half, while the waveform continues to be displayed properly.

### Code for reproduction

```python
import torchaudio
import torchaudio.transforms as transforms
import matplotlib.pyplot as plt
import numpy as np
import scipy.interpolate
import librosa.display
import matplotlib.gridspec as gridspec

from tqdm import tqdm
from concurrent.futures import ProcessPoolExecutor, as_completed


def compute_spectrogram_per_channel(channel_waveform, sample_rate):
    # Create transformer to convert waveform to spectrogram
    spectrogram_transform = transforms.Spectrogram(n_fft=2048, hop_length=256)

    # Apply the transformer
    spectrogram = spectrogram_transform(channel_waveform.unsqueeze(0))

    amplitude_spectrogram = np.sqrt(spectrogram)
    db_spectrogram = librosa.amplitude_to_db(amplitude_spectrogram[0].numpy(), ref=np.max)
    db_spectrogram = np.clip(db_spectrogram, a_min=None, a_max=0)  # clip to 0dB

    # Set new log scale
    num_freqs, num_frames = db_spectrogram.shape
    min_freq = 1  # human hearing range in Hz
    max_freq = sample_rate / 2  # Nyquist frequency
    frequencies = np.linspace(min_freq, max_freq, num=num_freqs)

    # Create a new scale
    log_scale = np.log10(frequencies)
    linear_scale = np.linspace(np.log10(min_freq), np.log10(max_freq), num=num_freqs)
    scale_ratio = 0.75  # adjust this parameter to control the ratio of log scale and linear scale
    new_scale = scale_ratio * log_scale + (1 - scale_ratio) * linear_scale

    new_db_spectrogram = np.empty_like(db_spectrogram)

    # Apply interpolation for each frame
    for frame in tqdm(range(num_frames)):
        interpolator = scipy.interpolate.interp1d(log_scale, db_spectrogram[:, frame])
        new_db_spectrogram[:, frame] = interpolator(new_scale)

    return channel_waveform.t().numpy(), db_spectrogram, new_db_spectrogram


def plot_spectrogram(waveforms, new_db_spectrograms, audio_duration):
    num_channels = len(waveforms)

    # Create a plot, set the background to black and adjust the size based on audio duration
    plt.figure(figsize=(max(audio_duration * 2, 10), 8), facecolor="black")

    # Dynamically create subplots based on the number of channels
    gs = gridspec.GridSpec(2 * num_channels, 1, height_ratios=[1] * num_channels + [5] * num_channels)

    # Loop through each channel to plot the waveform and spectrogram
    for i in range(num_channels):
        # Plot the waveform
        ax_waveform = plt.subplot(gs[i])
        ax_waveform.plot(waveforms[i], color="#4BF2A7")
        nonzero_indices = np.where(waveforms[i] != 0)[0]  # Find indices of non-zero values
        ax_waveform.set_xlim(nonzero_indices[0], nonzero_indices[-1])  # Set x limit to range of non-zero values
        ax_waveform.axis("off")

        # Plot the spectrogram
        ax_spectrogram = plt.subplot(gs[i + num_channels])
        ax_spectrogram.imshow(new_db_spectrograms[i], origin="lower", aspect="auto")
        ax_spectrogram.axis("off")

    plt.subplots_adjust(left=0, right=1, top=1, bottom=0, wspace=0, hspace=0)  # Adjust to remove borders and gaps
    plt.savefig("spectrogram.jpeg", facecolor="black", bbox_inches="tight", pad_inches=0)  # Save the figure


def main(file):
    # Load the audio file
    waveform, sample_rate = torchaudio.load(file)

    num_channels = waveform.shape[0]
    audio_duration = waveform.shape[1] / sample_rate  # Calculate audio duration

    with ProcessPoolExecutor() as executor:
        futures = {
            executor.submit(compute_spectrogram_per_channel, waveform[ch], sample_rate): ch
            for ch in range(num_channels)
        }
        waveforms = [None] * num_channels
        db_spectrograms = [None] * num_channels
        new_db_spectrograms = [None] * num_channels
        for future in as_completed(futures):
            ch = futures[future]
            waveforms[ch], db_spectrograms[ch], new_db_spectrograms[ch] = future.result()

    # Adjust figure width based on audio duration
    plot_spectrogram(waveforms, new_db_spectrograms, audio_duration)


# Run the main function
if __name__ == "__main__":
    file = "long.wav"
    main(file)
```

### Actual outcome

When visualizing an audio file of significant length, the spectrogram does not render correctly for the entire duration of the audio. The latter part of the spectrogram is blank and contains no information, which doesn't match with the waveform visualization, which continues to display normally.

long wave
spectrogram

short wave
spectrogram

### Expected outcome

The spectrogram should be consistently rendered for the entire duration of the audio file, matching the waveform visualization. Regardless of the length of the audio file, the spectrogram should be complete and not become blank at any point.

### Additional information

 - Other libraries involved: torchaudio, librosa, numpy, scipy

This issue occurs regardless of the audio file format or the specifics of the audio content. It seems directly related to the length of the audio file.

### Operating system

Windows & Ubuntu

### Matplotlib Version

3.7.2

### Matplotlib Backend

TkAgg

### Python version

Python 3.10.10

### Jupyter version

_No response_

### Installation

pip

### LINKED ISSUE #28893: [Bug]: Lines between points are invisible when there are more than 7 subfigures per row
### Bug summary

When there are more than 7 sub figures per row, the line between points will be invisible for the last several subfigs.

### Code for reproduction

```Python
import matplotlib.pyplot as plt

subfig_size = (16, 9)

num_subfigs = 10
fig, axes = plt.subplots(1, num_subfigs, figsize=(num_subfigs * subfig_size[0], 1 * subfig_size[1]), dpi=300)
for j in range(num_subfigs):
    xs = list(range(10))
    ax = axes[j]
    
    ax.plot(xs, xs, marker='x', label=f"test", linewidth=3, markersize=10)

plt.savefig('test.jpg')
```

### Actual outcome

test

### Expected outcome

Line should be visible!

### Additional information

What are the conditions under which this bug happens? input parameters, edge cases, etc?
* matplotlib==3.9.2
Has this worked in earlier versions?
* Tried 3.5.0, but still not works

### Operating system

Ubuntu

### Matplotlib Version

3.9.2

### Matplotlib Backend

qtagg

### Python version

3.10.9

### Jupyter version

6.5.2

### Installation

pip

### PR DESCRIPTION
**Pull request by QuLogic**
## PR summary

When rendering objects, Agg rasterizes them into scan line objects (an x/y point, horizontal length, and colour), and the renderer class writes those to the pixels in the final buffer. Though we have determined that Agg buffers cannot be larger than 2**16, the scan line classes that we use contain 16-bit _signed_ integers internally, cutting off positive values at half the maximum.

Since the renderer uses 32-bit integers, this can cause odd behaviour where any horizontal span that _starts_ before 2**15 (such as a horizontal spine) is rendered correctly even if it cross that point, but those that start after (such as a vertical spine or any portion of an angled line) end up clipped. For example, see how the spines and lines break in #28893.

image

A similar problem occurs for resampled images, which also uses Agg scanlines internally. See the breakage in #26368 for an example.

image

The example in that issue also contains horizontal spines that are wider than 2**15, which also exhibit strange behaviour.

image

By moving to 32-bit scan lines, positions and lengths of the lines will no longer be clipped, and this fixes rendering on very large figures.

Now, this does mean that internal bookkeeping will use double the memory. I have not yet benchmarked the memory usage to see what effect that would have. But I have implemented this two ways to show how it might work:
1. In `_image_resample.h`, the larger scan lines are used if the output image is >=2**15 in either dimension.
2. In `_backend_agg.h`, the larger scan lines are always used. While this is unfortunately the more common case, it's also the harder one to conditionalize as the type is in the class, and the template is used everywhere.

That being said, having the scan lines be conditional means that double the code is emitted instead. For the Agg backend, it increases by 4096 bytes (after stripping debug info), whereas for the image extension, it increased by 65536 bytes vs no increase for just using the larger one.

However, given the breakage in #23826, it may be necessary to _always_ use the 32-bit scan lines.

Fixes #23826
Fixes #26368
Fixes #28893

## PR checklist

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [ ] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [n/a] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [ ] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [ ] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

**Comment by tacaswell** [comment]
Even if we double the book keeping cost, I would expect it to be small in absolute terms?  Our output is small (on the scale of the data our users feed us, on the scale of a small of the footprint of a typical modern program and available memory on a typical machine).

**Comment by timhoffm** [comment]
... and given AGG is basically unchanged since 2006 and typical RAM sizes back then were about 512MB, I believe we can afford a worst-case factor of 2 increase now after 20 years.

**Comment by QuLogic** [comment]
I ran tests with `pytest-memray` (starting with #28856 to avoid the leak), and 99% of the tests didn't change peak memory usage. About 16 were within 50K, 3 more within 100K, and 17 over 1M bytes difference (both above and below the original). Running those 17 all over again, a lot were close to zero difference the second time over. I spot checked the remaining ones and the difference was in NumPy random state somewhere. So I think almost all the differences in peak memory usage were just global state and/or due to other tests running earlier depending on parallel job distribution.

File: src/_image_resample.h [hunk 1]
```diff
@@ -699,8 +702,8 @@ static void get_filter(const resample_params_t &params,
 }
 
 
-template<typename color_type>
-void resample(
+template<typename color_type, bool is_large = false>
```
Comment (timhoffm) [inline src/_image_resample.h:705]: Would this be a better name?
```suggestion
template<typename color_type, bool use_32bit = false>
```

**Review by timhoffm — APPROVED** [review]

File: src/_image_resample.h (hunk 1 above)
Comment (QuLogic) [reply src/_image_resample.h:705]: Since there's not really been much change to memory usage, I'm going to drop the template change and just always use the larger scan lines. That'll save us some code space, and template complications.

**Comment by QuLogic** [comment]
After some additional checks, I was able to determine that the image dimension limit is also outdated, and we can expand it to match Agg's coordinates at 2**23 pixels. Also after some searching, I believe I've essentially re-created #3451 which was closed due to the CXX removal.

With these changes, I am able to run:
```
import matplotlib.pyplot as plt
import numpy as np

t = np.arange(0, 3000, 0.01)
s1 = np.sin(2 * np.pi * 20 * t)

fig, ax = plt.subplots(figsize=(83886.07, .5), layout='constrained')
ax.plo(t, s1)
ax.set(xlim=(0, 2), xticks=[], yticks=[])
ax.grid(True)

fig.savefig('lines.png')
```
and produce an 8388607x50-pixel (i.e., 2**23-1) image with lines that are not broken in any visible way.

Shortening the example from #26368:
```
import numpy as np
import matplotlib.pyplot as plt

image = np.zeros((100000, 800, 3))
linspace = np.linspace(0, 1, 800)
image[:, :, 0] = 1 - linspace  # Red channel
image[:, :, 2] = linspace  # Blue channel

plt.figure(figsize=(10000, 0.5), dpi=100)
ax = plt.axes([0, 0, 1, 1], frameon=False)
ax.set_xticks([])
ax.set_yticks([])

ax.imshow(image, aspect="auto")

plt.savefig("gradient_image.png", pad_inches=0)
```
which produces a 1Mx50-pixel image of a gradient. Unfortunately, the resampling step takes about 16s per 100000 pixels, or, nearly 3 minutes for 1M pixels and 5-7GiB of RAM. So I did not test out the full 8M pixel image.

I also read issues that mentioned the 32768-pixel limit as being something about `sqrt(max int)`, i.e., total number of pixels below max int, though I'm not sure why. I believe we have changed most multiplications to upcast to `AGG_INT64`, and (2**23-1)**2 should be well within that. With RGBA, it's also approximately 256 TiB, so I have neither the RAM nor the time to test that.

**Comment by tacaswell** [comment]
Does this need a rebase?

**Comment by QuLogic** [comment]
It could to fix the CI, but from other PRs, it seems that macOS/homebrew broke again for other reasons.

**Review by greglucas — APPROVED** [review]
